pkg/apis/kudo/v1beta1/instance_validator.go

```diff
@@ -0,0 +1,36 @@
+package v1beta1
+
+import (
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/kudobuilder/kudo/pkg/util/kudo"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var _ kudo.Validator = &Instance{}
+
+// ValidateCreate implements webhookutil.validator (from controller-runtime)
+// we do not enforce any rules upon creation right now
+func (i *Instance) ValidateCreate(req admission.Request) error {
+	return nil
+}
+
+// ValidateUpdate hook called when UPDATE operation is triggered and our admission webhook is triggered
+// ValidateUpdate implements webhookutil.validator (from controller-runtime)
+func (i *Instance) ValidateUpdate(old runtime.Object, req admission.Request) error {
+	if i.Status.AggregatedStatus.Status.IsRunning() && req.RequestSubResource != "status" {
```

Comment:
I think the question is - how do you interrupt? I am not sure that interrupting by just overriding something in spec is the way to do. I would rather interrupt by explicitly triggering manual plan execution with some force parameter. Than all that will be handled via some extension on KEP-20 and all this will be ok...

Reply:
(I had an extensive comment here which I [unfortunately lost](https://github.com/alpha22jp/atomic-chrome/issues/52). I'll try to recreate it.)

In my experience one of the main sources of confusion for operators of SDK services are the implicit and opaque (sometimes undocumented) dependencies and conflicts between plans.

One thing that we'll need to think about (probably not in this PR) is how to deal with:
1. stuck plans (what is the behavior of starting a "deploy" plan when the previous "deploy" execution is stuck? for example, if a step can't complete because a container is crash-looping for some reason)
1. conflicting plans (what is the behavior of starting a "backup" plan when the "deploy" plan is either in progress or stuck on some step that could be seen as a dependency for any of the "backup" plan steps?)

For example, [this SDK test](https://github.com/mesosphere/dcos-commons/blob/ed6fbb09d5693b572dd16bbb7acb579ff95ad1aa/frameworks/helloworld/src/test/java/com/mesosphere/sdk/helloworld/scheduler/ServiceTest.java#L567-L735) shows a conflict scenario between the "deploy" and "recovery" plans in an SDK service.

The [plan execution section](https://mesosphere.github.io/dcos-commons/developer-guide/#plan-execution) of the SDK developer guide has some more (but not all) details about plan behavior.

Feedback from folks like @kaiwalyajoshi and @takirala might also be invaluable when thinking about improving plan behavior in the future and preventing past mistakes.

Reply:
Regarding interrupting plan executions, I think it does make sense. The SDK has
a concept of *pausing* plans. I think it'd be also useful to think about rolling
back plan executions.

From [the SDK developer guide](https://mesosphere.github.io/dcos-commons/developer-guide/#plan-execution):
> Normally, steps progress through statuses in the following order:

> PENDING → PREPARED → STARTING → COMPLETE

> The status of a phase or a plan is determined by examination of the step elements. A step may enter an ERROR state when its construction is malformed or whenever the service author determines it to be appropriate. The WAITING state occurs when the operator of the service indicates that an element should be paused. An operator might want to pause a deployment for a multitude of reasons, including unexpected failures during an update.